packages/adapter-firebase/src/index.ts

```diff
@@ -1,282 +1,236 @@
   getDoc,
-  getDocs,
-  getFirestore,
-  limit,
-  query,
-  runTransaction,
-  setDoc,
-  where,
-  connectFirestoreEmulator,
-} from "firebase/firestore"
-
-import type {
-  Adapter,
-  AdapterUser,
-  AdapterAccount,
-  AdapterSession,
-  VerificationToken,
-} from "next-auth/adapters"
-
-import { getConverter } from "./converter"
-
-export type IndexableObject = Record<string, unknown>
-
-export interface FirestoreAdapterOptions {
-  emulator?: {
-    host?: string
-    port?: number
-  }
+  getOneDoc,
+  mapFieldsFactory,
+} from "./utils"
+
+/** Configure the Firebase Adapter. */
+export interface FirebaseAdapterConfig extends AppOptions {
+  /**
+   * Use this option if mixed `snake_case` and `camelCase` field names in the database is an issue for you.
+   * Passing `snake_case` convert all field and collection names to `snake_case`.
+   * E.g. the collection `verificationTokens` will instead be `verification_tokens`,
+   * and fields like `emailVerified` will be `email_verified` instead.
+   *
+   * @default "default"
+   */
+  namingStrategy?: "snake_case" | "default"
 }
 
-export function FirestoreAdapter({
-  emulator,
-  ...firebaseOptions
-}: FirebaseOptions & FirestoreAdapterOptions): Adapter {
-  const firebaseApp = initializeApp(firebaseOptions)
-  const db = getFirestore(firebaseApp)
-
-  if (emulator) {
-    connectFirestoreEmulator(
-      db,
-      emulator?.host ?? "localhost",
-      emulator?.port ?? 3001
-    )
-  }
-
-  const Users = collection(db, "users").withConverter(
-    getConverter<AdapterUser & IndexableObject>()
-  )
-  const Sessions = collection(db, "sessions").withConverter(
-    getConverter<AdapterSession & IndexableObject>()
-  )
-  const Accounts = collection(db, "accounts").withConverter(
-    getConverter<AdapterAccount>()
-  )
-  const VerificationTokens = collection(db, "verificationTokens").withConverter(
-    getConverter<VerificationToken & IndexableObject>({ excludeId: true })
-  )
+/**
+ *
+ * #### Usage
+ *
+ * 1. Create a Firebase project and generate a service account key. `https://console.firebase.google.com/u/0/project/{project-id}/settings/serviceaccounts/adminsdk`
+ * 2. Download the service account key and save it in your project. (Make sure to add it to your `.gitignore`!)
+ * 3. Add `GOOGLE_APPLICATION_CREDENTIALS` to your environment variables and point it to the service account key file.
+ * 4. Add the adapter to your Auth.js / NextAuth.js configuration.
+ *
+ * ##### References
+ * - [Firebase Admin SDK setup](https://firebase.google.com/docs/admin/setup)
+ * - [`GOOGLE_APPLICATION_CREDENTIALS`](https://cloud.google.com/docs/authentication/application-default-credentials#GAC)
+ *
+ * ##### Example
+ *
+ * ```ts title="pages/api/auth/[...nextauth].ts"
+ * import NextAuth from "next-auth"
+ * import GoogleProvider from "next-auth/providers/google"
+ * import { FirestoreAdapter } from "@next-auth/firebase-adapter"
+ *
+ * export default NextAuth({
+ *   adapter: FirestoreAdapter(),
```

Comment:
Even if it's a bit more verbose, I still prefer forcing the user to pass a firestore instance themselves. I can only speak from personal experience, but I usually end up using the instance somewhere else on the backend. Also, I think it removes some guessing on how the credentials and appconfig are used to initialize the app. What do you think?

Reply:
Could support both, but it was something many people wanted. I think it depends - as you say - if you want to use the instance somewhere else. Will follow up.

Reply:
Thanks for handling both :)